File: src/Dignite.Sites.Domain/Seo/SeoConfiguration.cs
using Dignite.Abp.FlexFields;

namespace Dignite.Sites.Seo;

public static class SeoConfigurationNames
{
    public const string MetaTitleCharLimit = "Seo.MetaTitleCharLimit";
    public const string MetaDescriptionCharLimit = "Seo.MetaDescriptionCharLimit";
}

/// <summary>
/// Typed wrapper over <see cref="SeoFieldType"/>'s configuration (mirrors <c>TextConfiguration</c>).
/// <para>
/// Both limits are advisory, not enforced - <see cref="SeoFieldType.Validate"/> does not reject a longer
/// value, because a search engine truncates an over-length title or description rather than rejecting it.
/// They exist as a reference threshold for an editing UI or AI-generated content to aim for.
/// </para>
/// </summary>
public class SeoConfiguration : FieldConfigurationBase
{
    /// <summary>Default value: 60 - Google's practical truncation point for a search result title.</summary>
    public int MetaTitleCharLimit
    {
        get => ConfigurationDictionary.GetConfiguration(SeoConfigurationNames.MetaTitleCharLimit, 60);
        set => ConfigurationDictionary.SetConfiguration(SeoConfigurationNames.MetaTitleCharLimit, value);
    }

    /// <summary>Default value: 160 - the conventional SERP snippet length.</summary>
    public int MetaDescriptionCharLimit
    {
        get => ConfigurationDictionary.GetConfiguration(SeoConfigurationNames.MetaDescriptionCharLimit, 160);
        set => ConfigurationDictionary.SetConfiguration(SeoConfigurationNames.MetaDescriptionCharLimit, value);
    }

    public SeoConfiguration() : base()
    {
    }

    public SeoConfiguration(FieldConfigurationDictionary fieldConfiguration)
        : base(fieldConfiguration)
    {
    }
}

File: angular/projects/site/src/lib/field-types/seo/seo-configuration.ts
/**
 * Configuration of a `Seo` field, shaped for `FormBuilder.group()`. Mirrors `SeoConfiguration`
 * (`src/Dignite.Site.Domain/Seo/SeoConfiguration.cs`).
 *
 * The property names are the **stored** configuration keys, not a naming choice.
 *
 * Both limits are **advisory**. `SeoFieldType.Validate` does not reject an over-length value, because a
 * search engine truncates an over-long title or description rather than rejecting it. They are a
 * threshold for an editing UI (or AI-generated content) to aim for, so the control renders them as a
 * counter and never as a `Validators.maxLength`.
 */
export class SeoConfiguration {
  /** 60 - Google's practical truncation point for a search result title. */
  'Seo.MetaTitleCharLimit': unknown = [60];

  /** 160 - the conventional SERP snippet length. */
  'Seo.MetaDescriptionCharLimit': unknown = [160];
}
